examples/jax/encoder/test_multigpu_encoder.py

```diff
@@ -0,0 +1,398 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+""" Encoder training on multi-GPU with data parallelism"""
+import argparse
+import unittest
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+import nltk
+import numpy as np
+import optax
+import tensorflow_datasets as tfds
+from cuda import cudart
+from flax import linen as nn
+from flax.core.frozen_dict import FrozenDict
+from flax.training import train_state
+from jax.experimental import mesh_utils
+from jax.experimental.pjit import pjit
+
+import transformer_engine.jax as te
+
+DEVICE_DP_AXIS = 'data'
+PARAMS_KEY = 'params'
+PARAMS_AXES_KEY = PARAMS_KEY + '_axes'
+DROPOUT_KEY = 'dropout'
+INPUT_KEY = 'input_rng'
+NUM_GPU = 8
+
+
+def num_gpu_is_enough():
+    """Check if the machine has 8 GPUs."""
+    return len(jax.local_devices()) >= NUM_GPU
+
+
+def gpu_has_fp8():
+    """Check if the GPU has FP8."""
+    cudaSuccess = cudart.cudaError_t.cudaSuccess
+    ret, gpu_id = cudart.cudaGetDevice()
+    assert ret == cudaSuccess
+    flag = cudart.cudaDeviceAttr.cudaDevAttrComputeCapabilityMajor
+    _, major = cudart.cudaDeviceGetAttribute(flag, gpu_id)
+    flag = cudart.cudaDeviceAttr.cudaDevAttrComputeCapabilityMinor
+    _, minor = cudart.cudaDeviceGetAttribute(flag, gpu_id)
+    sm_arch = major * 10 + minor
+    return sm_arch >= 89
+
+
+class Net(nn.Module):
+    """NLP Encoder"""
+    num_embed: int
+
+    @nn.compact
+    def __call__(self, x, disable_dropout=False):
+        x = nn.Embed(num_embeddings=self.num_embed, features=256, dtype=jnp.bfloat16)(x)
+
+        te_Encoder = partial(te.TransformerLayer,
+                             hidden_size=256,
+                             mlp_hidden_size=1024,
+                             num_attention_heads=8,
+                             hidden_dropout=0.1,
+                             attention_dropout=0.1,
+                             dropout_rng_name=DROPOUT_KEY,
+                             layer_type=te.TransformerLayerType.ENCODER,
+                             enable_relative_embedding=False,
+                             dtype=jnp.bfloat16)
+        x = te_Encoder()(x, deterministic=disable_dropout)
+
+        x = x.reshape(x.shape[0], -1)
+
+        x = te.DenseGeneral(features=256, sharding_type=te.ShardingType.DP, dtype=jnp.bfloat16)(x)
+
+        x = te.DenseGeneral(features=256, sharding_type=te.ShardingType.DP, dtype=jnp.bfloat16)(x)
+
+        x = nn.Dense(features=2, dtype=jnp.bfloat16)(x)
+        return x
+
+
+def train_step(state, inputs, labels, var_collect, rngs, use_fp8):
```

Comment:
This example uses `pjit` to set up complex `PartitionSpec` instead of `jax.jit`
```python
            in_shardings = (state_pspec, inputs_pspec, labels_pspec, None, None)
            out_shardings = (state_pspec, None, None, None)
            pjit_train_step = pjit(train_step, in_shardings, out_shardings, static_argnums=(5,))
```